[WIP] geo: prototype library for spatial data operations

danhhz opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45523)
<!-- Reviewable:end -->

danhhz commented:
There's already a few things I'd like to change about this prototype, but I wanted to get it out there early for discussion and input. (Also, in case it's helpful to otan for his SQL prototyping.)

otan commented on pkg/geo/geo.go:
```diff
@@ -0,0 +1,75 @@
+// Region represents a geometrical object in planar or spherical space.
+//
+// WIP This needs a better name. GeoJSON calls it a "Geometry", but that's
+// confusing with the SQL GEOMETRY type. Does PostGIS have a name for this? A
```
postgis seems to call the data types `GEOMETRY` and `GEOGRAPHY`. they seem to both be serialized the same as a GSERIALIZED (http://postgis.net/docs/doxygen/2.5/df/d2d/structGSERIALIZED.html -- https://github.com/postgis/postgis/blob/master/liblwgeom/gserialized.txt), so maybe they can keep the same structure.

geography in returns `gserialized`: https://sourcegraph.com/github.com/postgis/postgis/-/blob/postgis/geography_inout.c#L136

geometry uses `LWGEOM_in` (https://sourcegraph.com/github.com/postgis/postgis/-/blob/postgis/postgis.sql.in#L113:24): https://sourcegraph.com/github.com/postgis/postgis/-/blob/postgis/lwgeom_inout.c#L84:1 which returns `GSERIALIZED` as well, so they are the same internally.

`Geography` seems to have a stricter set of types allowed in but that's about it.

geography can be explicitly cast to geometry, but geometry can be implicitly cast to geography

otan commented on pkg/geo/geo.go:
```diff
@@ -0,0 +1,75 @@
+	s2.Region
+	// TODO(dan): Most operations on geometrical objects need to have SRIDs
+	// associated with them to either do the conversion or error. Where should
+	// those SRIDs live? Here?
```
it appears so, if we were following http://postgis.net/docs/doxygen/2.5/df/d2d/structGSERIALIZED.html

danhhz replied:
TIL this is in the spec too:

> The OpenGIS specification also requires that the internal storage format of spatial objects include a spatial referencing system identifier (SRID). The SRID is required when creating spatial objects for insertion into the database.

otan commented on pkg/geo/geo.go:
```diff
@@ -0,0 +1,75 @@
+	// WIP The prototype has made it pretty clear that using s2.Region here
+	// results in awkward type switches all over the place. There's got to be
+	// something better for us to do here.
+	s2.Region
```
wondering if we should make `s2.Region` the thing to inherit, or to make our own struct hierarchy that implements the inheritance properly. using `oneof` a la protobufs comes to mind....

danhhz committed:
vendor: add new geo deps

Resolving vendor merge conflicts is costly, so get these in early.

Release note: None

danhhz committed:
geo: begin a spatial library

Included:
- A roundtrip-able long term storage format (implemented for a small
  number of the shapes we'll support)
- Conversion between GEOMETRY and GEOGRAPHY
- The very beginnings of SRID support

Release note: None

danhhz commented:
Moving this to a feature branch for now, so Oliver and I can collaborate on our prototypeing
